File: LeetCode/remote/word_break.cs
//  https://leetcode.com/problems/word-break/
//
//   Given a string s and a dictionary of words dict,
//   determine if s can be segmented into a space-separated sequence of one or more dictionary words.
//
//   For example, given
//   s = "leetcode",
//   dict = ["leet", "code"].
//
//   Return true because "leetcode" can be segmented as "leet code". 

using System;
using System.Collections.Generic;
using System.Linq;

public class Solution {
    //  
    //  Submission Details
    //  32 / 32 test cases passed.
    //      Status: Accepted
    //      Runtime: 184 ms
    //          
    //          Submitted: 0 minutes ago
    //
    //  https://leetcode.com/submissions/detail/50902355/
    //
    public bool WordBreak(string s, ISet<string> wordDict) {
        var n = s.Length;
        var wb = new bool[n + 1];
        wb[0] = true;
        for (var i = 0; i < n; i++)
        {
            if (!wb[i])
            {
                continue;
            }

            var current = s.Substring(i);
            foreach (var word in wordDict)
            {
                if (current.StartsWith(word))
                {
                    wb[i + word.Length] = true;
                }
            }
        }

        return wb[n];
    }

    public bool WordBreakRec(string s, ISet<string> wordDict) {
        if (String.IsNullOrEmpty(s))
        {
            return true;
        }

        return wordDict.
               Aggregate(
                    false,
                    (acc, x) => acc || s.StartsWith(x) && WordBreakRec(s.Substring(x.Length), wordDict));
    }

    //  https://leetcode.com/submissions/detail/61354399/
    //
    //  Submission Details
    //  34 / 34 test cases passed.
    //      Status: Accepted
    //      Runtime: 196 ms
    //          
    //          Submitted: 0 minutes ago
    //
    public bool WordBreak(string s, ISet<string> wordDict, Dictionary<string, bool> broken = null) {
        if (broken == null)
        {
            broken = new Dictionary<string, bool>();
        }
        
        if (broken.ContainsKey(s))
        {
            return broken[s];
        }
        
        if (String.IsNullOrWhiteSpace(s))
        {
            return true;
        }
        
        return broken[s] = wordDict
                           .Where(x => s.StartsWith(x))
                           .Select(x => WordBreak(s.Substring(x.Length), wordDict, broken))
                           .Any(x => x);
    }

    //  https://leetcode.com/submissions/detail/63800796/
    //
    //  Submission Details
    //  34 / 34 test cases passed.
    //      Status: Accepted
    //      Runtime: 196 ms
    //          
    //          Submitted: 1 minute ago
    //
    public bool WordBreak(string s, ISet<string> wordDict, Dictionary<String, bool> hash = null) {
        if (String.IsNullOrWhiteSpace(s)) {
            return true;
        }
        
        if (hash == null) {
            hash = new Dictionary<String, bool>();
        }
        
        if (!hash.ContainsKey(s)) {
            hash[s] = wordDict //   And the below boolean statement is exactly why we can't have nice things
                      .Any(x => !s.StartsWith(x) ? false : WordBreak(s.Substring(x.Length), wordDict, hash));
        }
        
        return hash[s];
    }

    static void Main()
    {
        var s = new Solution();
        Console.WriteLine(s.WordBreakRec("leetcode", new HashSet<String> { "leet", "code" }));
        Console.WriteLine(s.WordBreak("leetcode", new HashSet<String> { "leet", "code" }));
        Console.WriteLine(s.WordBreakRec("leetcode", new HashSet<String> { "leet", "cod" }));
        Console.WriteLine(s.WordBreak("leetcode", new HashSet<String> { "leet", "cod" }));
    }
}

File: LeetCode/remote/word_break.rb
#   https://leetcode.com/problems/word-break/
#
#    Given a string s and a dictionary of words dict, determine if s can be segmented into a space-separated sequence of one or more dictionary words. 

# TLE
def word_break(s, word_dict)
    return s.to_s.empty? || 
           (1..s.size)
           .select { |x| word_dict.include?(s[0, x]) }
           .map { |x| word_break(s[x..-1], word_dict) }
           .inject(:|)
end

# Still TLE
def word_break(s, word_dict, hash = nil)
    return hash[s] if !hash.nil? && !hash[s].nil?
    hash ||= {}
    result = s.to_s.empty? || 
             (1..s.size)
             .select { |x| word_dict.include?(s[0, x]) }
             .map { |x| word_break(s[x..-1], word_dict, hash) }
             .inject(:|)
    hash[s] = result || false
    return hash[s]
end

#   https://leetcode.com/submissions/detail/54508959/
#
#   Submission Details
#   34 / 34 test cases passed.
#       Status: Accepted
#       Runtime: 128 ms
#           
#           Submitted: 0 minutes ago

def word_break(s, word_dict)
    wb = Array.new(s.size + 1) { false }
    wb[0] = true
    0.upto(s.size - 1).each { |i|
        next if !wb[i]
        1.upto(s.size - i).each { |j|
            next if wb[i + j]
            wb[i + j] = word_dict.include?(s[i, j])
        }
    }

    wb[s.size]
end

# Small optimization (old idea of mine on the C# solution)
# to push this solution to the next level
#
# https://leetcode.com/submissions/detail/54509068/
#
# Submission Details
# 34 / 34 test cases passed.
#   Status: Accepted
#   Runtime: 104 ms
#       
#       Submitted: 0 minutes ago
# You are here!
# Your runtime beats 100.00% of rubysubmissions.

def word_break(s, word_dict)
    wb = Array.new(s.size + 1) { false }
    wb[0] = true
    0.upto(s.size - 1).each { |i|
        next if !wb[i]
        current = s[i..-1]
        word_dict.select { |word| current.start_with?(word) }.each { |word|
            wb[i + word.size] = true
        }
    }

    wb[s.size]
end
